### DISCUSSION TITLE
How to Allow YOLOv5 to Draw Multiple Overlapping Bounding Boxes

### DISCUSSION THREAD

**Discussion by mermetal** [ultralytics/ultralytics#12789] [Q&A]
I understand that NMS (Non-Maximum Suppression) works to draw only the bounding box with the highest confidence when a specific IoU value is met. 
However, I want to apply this function only "within the same class." For example, when detecting an image of a soccer player dribbling a ball, I want to draw a 'dribble' bounding box that includes both the ball and the dribbling player, and at the same time, draw a 'ball' bounding box around the ball and a 'person' bounding box around the player within the 'dribble' bounding box.
In this case, only one bounding box is drawn for each class, but ultimately, bounding boxes for objects of different classes will overlap.

Which part of the non_max_suppression function should I modify to achieve this?

**Comment by glenn-jocher** [comment]
@mermetal to allow YOLOv5 to draw multiple overlapping bounding boxes for different classes while performing class-specific Non-Maximum Suppression (NMS), you should modify the `non_max_suppression` function to handle suppression separately for each class. Here's a simple way you can adjust the existing function:

1. Ensure that the suppression is done per class by grouping detections by their classes before applying NMS.
2. You can modify the section of the code in `utils/general.py` where NMS is applied. Look for the loop that processes each detection and include a step to separate detections by class.

Here's a brief code snippet that demonstrates this idea:

```python
import torch

def non_max_suppression(prediction, conf_thresh=0.25, iou_thresh=0.45, classes=None, agnostic=False):
    # This is a simplified version - adjust according to your specific needs
    bs = prediction.shape[0]
    nc = prediction.shape[2] - 5  # number of classes
    xc = prediction[..., 4] > conf_thresh  # candidates

    # Settings for each image
    output = [None] * bs
    for xi, x in enumerate(prediction):  # image index, image predictions
        x = x[xc[xi]]  # confidence

        # Check classes
        if classes:
            x = x[(x[:, 5:nc] == torch.tensor(classes, device=x.device)).any(1)]

        # Class-wise NMS
        box = x[:, :4].clone()
        conf = x[:, 4].clone()
        for c in range(nc):
            if agnostic:
                pass  # skip class-agnostic NMS
            else:
                c_indices = (x[:, 5 + c] == 1).nonzero(as_tuple=False).flatten()
                x_class = x[c_indices]
                nms_indices = torchvision.ops.nms(x_class[:, :4], x_class[:, 4], iou_thresh)
                output[xi] = x_class[nms_indices]

    return output
```

📘 Note: Ensure to test this modification to verify it behaves as expected in your scenario, specifically checking the interactions near the class assignments and suppression conditions.

Keep iterating based on your needs and dataset! 🚀
